core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -257,22 +257,48 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
     val partitionsByDestination: immutable.Map[Node, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
       var brokerNode: Option[Node] = None
 
-      // TODO: instead of retry until succeed, we can first put it into an unknown broker queue and let the sender thread to look for its broker and migrate them
-      while (brokerNode.isEmpty) {
-        brokerNode = metadataCache.getPartitionLeaderEndpoint(topicPartition.topic, topicPartition.partition, interBrokerListenerName)
+      brokerNode = metadataCache.getPartitionLeaderEndpoint(topicPartition.topic, topicPartition.partition, interBrokerListenerName)
```

Comment:
Hmm. I guess I need to sync face to face. But from the code it seems like if there is no leader for any single topic partition, then the operation for the entire transactionalId in the purgatory will be marked as completed? If this is true, then I have two questions: 

 1. Would we enter this case if there are no live replicas for a particular partition? Or will `metadataCache.getPartitionLeaderEndpoint` only return no brokers if the topic is deleted? 
2. Assuming that no leader means the partition is deleted, would the current logic mark the operation in purgatory as successful if even one partition in the transaction was deleted? 

As I said, we can discuss this face to face: it may be more efficient as I am not most familiar with a bunch of this stuff.